src/relay/ir/hash.cc

```diff
@@ -0,0 +1,318 @@
+      return found->second;
+    } else {
+      auto hash = this->VisitType(type);
+      hash_map_.insert({type, hash});
+      return hash;
+    }
+  }
+  /*!
+   * Compute the hash of an expression.
+   *
+   * \note We run graph structural equality checking when comparing two Exprs.
+   *   This means that AlphaEqualHandler can only be used once for each pair.
+   *   The equality checker checks data-flow equvalence of the Expr DAG.
+   *   This function also runs faster as it memomizes equal_map.
+   *
+   * \param expr The expression to hash.
+   * \return the hash value.
+   */
+  size_t ExprHash(const Expr& expr) {
+    if (!expr.defined()) return expr.hash();
+    auto found = hash_map_.find(expr);
+    if (found != hash_map_.end()) {
+      return found->second;
+    } else {
+      auto hash = this->VisitExpr(expr);
+      hash_map_.insert({expr, hash});
+      return hash;
+    }
+  }
+
+ protected:
+  /*!
+   * \brief Hash a DataType.
+   * \param dtype The dtype to hash.
+   * \return the hash value.
+   */
+  size_t DataTypeHash(const DataType& dtype) {
+     return std::hash<int>()(
+         static_cast<int>(dtype.code()) |
+         (static_cast<int>(dtype.bits()) << 8) |
+         (static_cast<int>(dtype.lanes()) << 16));
+  }
+
+  /*!
+   * \brief Check Equality of leaf node of the graph.
+   *  if map_free_var_ is set to true, try to map via equal node.
+   * \param lhs The left hand operand.
+   * \param rhs The right hand operand.
+   * \return the compare result.
+   */
+  size_t LeafNodeEqual(const NodeRef& lhs, const NodeRef& rhs) {
+    return 0;
+    // if (lhs.same_as(rhs)) return true;
+    // auto it = equal_map_.find(lhs);
+    // if (it != equal_map_.end()) {
+    //   return it->second.same_as(rhs);
+    // } else {
+    //   if (map_free_var_) {
+    //     if (lhs->type_index() != rhs->type_index()) return false;
+    //     equal_map_[lhs] = rhs;
+    //     return true;
+    //   } else {
+    //     return false;
+    //   }
+    // }
+  }
+
+  using AttrsHashHandler::VisitAttr_;
+  size_t VisitAttr_(const Variable* lhs) final {
+    return 0; // return LeafNodeEqual(GetRef<NodeRef>(lhs), other);
+  }
+
+  // Type hashing
+  size_t VisitType_(const TensorTypeNode* tensor_type) final {
+    size_t hash = std::hash<std::string>()(tensor_type->_type_key);
+    hash = Combine(hash, DataTypeHash(tensor_type->dtype));
+    hash = Combine(hash, Hash(tensor_type->shape));
+    return hash;
+  }
+
+  size_t VisitType_(const IncompleteTypeNode* incomplete) final {
+    return GetRef<IncompleteType>(incomplete);
+  }
+
+  size_t VisitType_(const TypeVarNode* tyvar) final {
+    int index = BindVar(GetRef<TypeVar>(tyvar));
+    return std::hash<int>()(index);
+  }
+
+  size_t VisitType_(const FuncTypeNode* func_type) final {
+    size_t hash = std::hash<std::string>()(func_type->_type_key);
+    for (auto type_param : func_type->type_params) {
+      hash = Combine(hash, TypeHash(type_param));
```

Comment:
This is a declaration side, need to record the hash value of type_param into the hash_map_